Zotero/Scenes/Detail/PDF/ViewModels/PDFReaderActionHandler.swift

```diff
@@ -1780,7 +1782,26 @@ final class PDFReaderActionHandler: ViewModelActionHandler, BackgroundDbProcessi
             guard let annotation = notification.object as? PSPDFKit.Annotation else { return }
 
             if let changes = notification.userInfo?[PSPDFAnnotationChangedNotificationKeyPathKey] as? [String] {
-                self.change(annotation: annotation, with: changes, in: viewModel)
+                if let freeTextAnnotation = annotation as? PSPDFKit.FreeTextAnnotation, let key = annotation.key {
+                    if changes.contains("rotation") {
+                        // Debounce these notifications because FreeTextAnnotation rotation change spams these annotations in milliseconds
+                        // and it looks bad in sidebar while it's also unnecessary cpu burden.
+                        let disposeBag = DisposeBag()
+                        freeTextAnnotationRotationDebounceDisposeBagByKey[key] = disposeBag
+                        Single<Int>.timer(.milliseconds(100), scheduler: MainScheduler.instance)
+                            .subscribe(onSuccess: { [weak self, weak viewModel] _ in
+                                guard let self, let viewModel else { return }
+                                change(annotation: annotation, with: changes, in: viewModel)
+                                freeTextAnnotationRotationDebounceDisposeBagByKey[key] = nil
+                            })
+                            .disposed(by: disposeBag)
+                    } else {
+                        change(annotation: annotation, with: changes, in: viewModel)
+                        freeTextAnnotationRotationDebounceDisposeBagByKey[key] = nil
```

Comment:
Wouldn't this cancel previous rotation notification? You can leave `freeTextAnnotationRotationDebounceDisposeBagByKey` alone here, it'll be reset when it's debounced.

Reply:
It's change for the same annotation, but the change doesn't contain `rotation` change this time. The `func change(annotation: PSPDFKit.Annotation, with changes: [String], in viewModel: ViewModel<PDFReaderActionHandler>)` checks `changes` array to see which changes were applied and applies only those changes.

So if for example you got a `rotation` change and then suddenly a `color` change, rotation wouldn't be applied, only new color. I know that the 100 ms debounce won't really let this happen, but we may increase the debounce later while not being aware of this.

Reply:
New commit should work fine, feel free to merge

Reply:
I misunderstood that. Check https://github.com/zotero/zotero-ios/pull/943/commits/2328bf63dc8bf3e8019dd244e0f9877528364e51 for a solution that maintains proper order of execution.